From 8408716a7f2909eab4618ac2ab6e9a1565a993d0 Mon Sep 17 00:00:00 2001
From: Mark Thomas <markt@apache.org>
Date: Thu, 3 Nov 2016 09:38:09 +0000
Subject: [PATCH] Fix DBCP-457

When using a BasicDataSource, pass changes related to the handling of
abandoned connections to the underlying pool so that the pool
configuration may be updated dynamically.
---
 src/changes/changes.xml                       |  5 ++++
 .../apache/commons/dbcp2/BasicDataSource.java | 24 +++++++++++++++++++
 .../commons/dbcp2/TestBasicDataSource.java    | 21 ++++++++++++++++
 3 files changed, 50 insertions(+)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index c1b5d68bcb..0fc7a0eba1 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -122,6 +122,11 @@ The <action> type attribute can be add,update,fix,remove.
          Make it simpler to extend BasicDataSource to allow sub-classes to
          provide custom GenericObjectPool implementations.
        </action>
+       <action dev="markt" type="fix" issue="DBCP-457">
+         When using a BasicDataSource, pass changes related to the handling of
+         abandoned connections to the underlying pool so that the pool
+         configuration may be updated dynamically. 
+       </action>
     </release>
     <release version="2.1.1" date="6 Aug 2015" description=
 "This is a patch release, including bug fixes only.">
diff --git a/src/main/java/org/apache/commons/dbcp2/BasicDataSource.java b/src/main/java/org/apache/commons/dbcp2/BasicDataSource.java
index 250dfd9220..12314c8090 100644
--- a/src/main/java/org/apache/commons/dbcp2/BasicDataSource.java
+++ b/src/main/java/org/apache/commons/dbcp2/BasicDataSource.java
@@ -1667,6 +1667,10 @@ public void setRemoveAbandonedOnMaintenance(
         }
         abandonedConfig.setRemoveAbandonedOnMaintenance(
                 removeAbandonedOnMaintenance);
+        GenericObjectPool<?> gop = this.connectionPool;
+        if (gop != null) {
+            gop.setAbandonedConfig(abandonedConfig);
+        }
     }
 
     /**
@@ -1699,6 +1703,10 @@ public void setRemoveAbandonedOnBorrow(final boolean removeAbandonedOnBorrow) {
             abandonedConfig = new AbandonedConfig();
         }
         abandonedConfig.setRemoveAbandonedOnBorrow(removeAbandonedOnBorrow);
+        GenericObjectPool<?> gop = this.connectionPool;
+        if (gop != null) {
+            gop.setAbandonedConfig(abandonedConfig);
+        }
     }
 
     /**
@@ -1744,6 +1752,10 @@ public void setRemoveAbandonedTimeout(final int removeAbandonedTimeout) {
             abandonedConfig = new AbandonedConfig();
         }
         abandonedConfig.setRemoveAbandonedTimeout(removeAbandonedTimeout);
+        GenericObjectPool<?> gop = this.connectionPool;
+        if (gop != null) {
+            gop.setAbandonedConfig(abandonedConfig);
+        }
     }
 
     /**
@@ -1772,6 +1784,10 @@ public void setLogAbandoned(final boolean logAbandoned) {
             abandonedConfig = new AbandonedConfig();
         }
         abandonedConfig.setLogAbandoned(logAbandoned);
+        GenericObjectPool<?> gop = this.connectionPool;
+        if (gop != null) {
+            gop.setAbandonedConfig(abandonedConfig);
+        }
     }
 
     /**
@@ -1796,6 +1812,10 @@ public void setAbandonedLogWriter(final PrintWriter logWriter) {
             abandonedConfig = new AbandonedConfig();
         }
         abandonedConfig.setLogWriter(logWriter);
+        GenericObjectPool<?> gop = this.connectionPool;
+        if (gop != null) {
+            gop.setAbandonedConfig(abandonedConfig);
+        }
     }
 
     /**
@@ -1829,6 +1849,10 @@ public void setAbandonedUsageTracking(final boolean usageTracking) {
             abandonedConfig = new AbandonedConfig();
         }
         abandonedConfig.setUseUsageTracking(usageTracking);
+        GenericObjectPool<?> gop = this.connectionPool;
+        if (gop != null) {
+            gop.setAbandonedConfig(abandonedConfig);
+        }
     }
 
     // --------------------------------------------------------- Public Methods
diff --git a/src/test/java/org/apache/commons/dbcp2/TestBasicDataSource.java b/src/test/java/org/apache/commons/dbcp2/TestBasicDataSource.java
index 3f1667db9e..db99049081 100644
--- a/src/test/java/org/apache/commons/dbcp2/TestBasicDataSource.java
+++ b/src/test/java/org/apache/commons/dbcp2/TestBasicDataSource.java
@@ -35,6 +35,7 @@
 
 import org.apache.commons.logging.LogFactory;
 import org.junit.After;
+import org.junit.Assert;
 import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Test;
@@ -795,6 +796,26 @@ public void testDisconnectSqlCodes() throws Exception {
         assertTrue(pcf.getDisconnectionSqlCodes().contains("XXX"));
         assertEquals(1, pcf.getDisconnectionSqlCodes().size());
     }
+
+    /**
+     * JIRA: DBCP-457
+     * Verify that changes made to abandoned config are passed to the underlying
+     * pool.
+     */
+    @Test
+    public void testMutateAbandonedConfig() throws Exception {
+        final Properties properties = new Properties();
+        properties.put("initialSize", "1");
+        properties.put("driverClassName", "org.apache.commons.dbcp2.TesterDriver");
+        properties.put("url", "jdbc:apache:commons:testdriver");
+        properties.put("username", "foo");
+        properties.put("password", "bar");
+        final BasicDataSource ds = BasicDataSourceFactory.createDataSource(properties);
+        boolean original = ds.getConnectionPool().getLogAbandoned();
+        ds.setLogAbandoned(!original);
+        Assert.assertNotEquals(Boolean.valueOf(original),
+                Boolean.valueOf(ds.getConnectionPool().getLogAbandoned()));
+    }
 }
 
 /**
